File: 0133-clone-graph/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if(node == null) return null;
        if(node.neighbors.isEmpty()) return new Node(node.val);

        Queue<Node> q = new LinkedList<>();
        HashMap<Node, Node> copy = new HashMap<>();

        q.offer(node);
        copy.put(node, new Node(node.val));

        while(!q.isEmpty()){
            Node curr = q.poll();

            for(Node neigh: curr.neighbors){
                if(!copy.containsKey(neigh)){
                    copy.put(neigh, new Node(neigh.val));
                    q.add(neigh);
                }
                copy.get(curr).neighbors.add(copy.get(neigh));
            }
        }

        return copy.get(node);

    }
}

File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution(object):
    def cloneGraph(self, node):
        """
        :type node: Node
        :rtype: Node
        """
        if not node:
            return None 

        if not node.neighbors:
            return Node(node.val)

        q = deque()
        q.append(node)

        copy = {}
        copy[node] = Node(node.val)

        while q:
            size = len(q)

            for i in range(0, size):
                curr = q.popleft()

                for next in curr.neighbors:
                    if next not in copy:
                        copy[next] = Node(next.val)
                        q.append(next)
                    copy[curr].neighbors.append(copy[next])

        return copy[node]
                    

        